src/recipes.zig

```diff
@@ -0,0 +1,234 @@
+			if(keys.get(symbol)) |literal| {
+				if(segments.items.len > 0 and segments.items[segments.items.len - 1] == .literal) {
+					const value = segments.pop();
+					defer allocator.free(value.literal);
+					segments.append(.{.literal = std.mem.concat(allocator.allocator, u8, &.{value.literal, literal}) catch unreachable});
+				} else {
+					segments.append(.{.literal = allocator.dupe(u8, literal)});
+				}
+			} else {
+				segments.append(.{.symbol = allocator.dupe(u8, symbol)});
+			}
+			idx = endIndex + 1;
+		} else {
+			const endIndex = std.mem.indexOfScalarPos(u8, pattern, idx, '{') orelse pattern.len;
+			if(segments.items.len > 0 and segments.items[segments.items.len - 1] == .literal) {
+				const value = segments.pop();
+				defer allocator.free(value.literal);
+				segments.append(.{.literal = std.mem.concat(allocator.allocator, u8, &.{value.literal, pattern[idx..endIndex]}) catch unreachable});
+			} else {
+				segments.append(.{.literal = allocator.dupe(u8, pattern[idx..endIndex])});
+			}
+			idx = endIndex;
+		}
+	}
+	return segments;
+}
+
+fn matchWithKeys(allocator: NeverFailingAllocator, target: []const u8, pattern: []const Segment, keys: *const std.StringHashMap([]const u8)) !std.StringHashMap([]const u8) {
+	var idx: usize = 0;
+	idx = 0;
+	var newKeys = keys.clone() catch unreachable;
+	errdefer newKeys.deinit();
+	for(0.., pattern) |i, segment| {
+		switch(segment) {
+			.literal => |literal| {
+				if(literal.len + idx > target.len or !std.mem.eql(u8, target[idx .. idx + literal.len], literal)) {
+					return error.NoMatch;
+				}
+				idx += literal.len;
+			},
+			.symbol => |symbol| {
+				const endIndex: usize = if(i + 1 < pattern.len) blk: {
+					const nextSegment = pattern[i + 1];
+					if(nextSegment == .symbol) {
+						return error.NoMatch;
+					}
+					break :blk std.mem.indexOfPos(u8, target, idx, nextSegment.literal) orelse {
+						return error.NoMatch;
+					};
+				} else target.len;
+				newKeys.put(allocator.dupe(u8, symbol), allocator.dupe(u8, target[idx..endIndex])) catch unreachable;
+				idx = endIndex;
+			},
+		}
+	}
+	if(idx == target.len) {
+		return newKeys;
+	} else {
+		return error.NoMatch;
+	}
+}
+
+const ItemKeyPair = struct {item: ItemStack, keys: std.StringHashMap([]const u8)};
+
+fn parseRecipeItem(allocator: NeverFailingAllocator, zon: ZonElement, keys: *const std.StringHashMap([]const u8)) !main.List(ItemKeyPair) {
+	var id = zon.as([]const u8, "");
+	id = std.mem.trim(u8, id, &std.ascii.whitespace);
+	var amount: u16 = 1;
+	if(std.mem.indexOfScalar(u8, id, ' ')) |index| blk: {
+		amount = std.fmt.parseInt(u16, id[0..index], 0) catch break :blk;
+		id = id[index + 1 ..];
+		id = std.mem.trim(u8, id, &std.ascii.whitespace);
+	}
+	var itemPairs: main.List(ItemKeyPair) = .initCapacity(allocator, 1);
+	var iterator = std.mem.splitScalar(u8, id, '.');
+	id = iterator.next().?;
+	var tags: main.List(Tag) = .init(allocator);
+	defer tags.deinit();
```

Comment:
Please don't use the passed in allocator for purely local allocations. I'd suggest to also use another local arena in this function, especially given the long defer block below.